### PR TITLE
feat(pack): support `publishConfig.type`

### PR DESCRIPTION
**Pull request by cometkim**
**What's the problem this PR addresses?**

This PR allows overriding the `"type"` field on the manifest when packing a package.

```json
{
  "main": "./src/index.ts",
  "publishConfig": {
    "type": "module",
    "main": "./lib/index.js"
  }
}
```
I wanna build an ESM-first package, but at the same time, I still wanna implicitly load packages with legacy Node resolution in my dev environment to interop with TypeScript which don't yet fully support Node ESM resolution

...

**How did you fix it?**

by supporting overriding `"type"` with `"publishConfig"` object

...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/gatsby/static/configuration/manifest.json [hunk 1]
```diff
@@ -288,6 +288,10 @@
           "enum": ["public", "restricted"],
           "examples": ["public"]
         },
+        "type": {
+          "$ref": "#/properties/type",
+          "description": "Same principle as the `publishConfig.type` property; this value will be used instead of the top-level `type` field when generating the workspace tarball."
```
Comment (merceyz) [inline packages/gatsby/static/configuration/manifest.json:293]: That's a reference to itself (self-reference)

**Review by merceyz — CHANGES_REQUESTED** [review]
Could you add a test to https://github.com/yarnpkg/berry/blob/f2513070530f4d309e3b398170a0d13865c104a8/packages/acceptance-tests/pkg-tests-specs/sources/commands/pack.test.js?

File: packages/gatsby/static/configuration/manifest.json (hunk 1 above)
Comment (cometkim) [reply packages/gatsby/static/configuration/manifest.json:293]: I think that's a reference to https://github.com/yarnpkg/berry/pull/4204/files#diff-b98641d8f09abf73c08b54295fb7e4fbd50cd4ea2275b602e4659de318865433R32-R37

**Review by arcanis — COMMENTED** [review]
I think the right field is `types` (or `typings`, but better only support one), not `type` 🤔

File: packages/gatsby/static/configuration/manifest.json (hunk 1 above)
Comment (merceyz) [reply packages/gatsby/static/configuration/manifest.json]: The description says that `publishConfig.type` is the "same principle as `publishConfig.type`" which is a self-reference and doesn't say much. I'm guessing you meant to point to `publishConfig.bin`

**Comment by merceyz** [comment]
@arcanis `type` is the correct field in this case https://nodejs.org/dist/latest-v16.x/docs/api/packages.html#type

**Comment by arcanis** [comment]
Of course, my bad 🤦‍♀️

File: packages/gatsby/static/configuration/manifest.json (hunk 1 above)
Comment (cometkim) [reply packages/gatsby/static/configuration/manifest.json]: Uh oh, yes yes I see

**Review by merceyz — APPROVED** [review]

**Comment by arcanis** [comment]
Sorry for the delay @cometkim 🙏

**Pull request merged**